sematic/external_resource.py

```diff
@@ -0,0 +1,229 @@
+        if not isinstance(self.status, ResourceStatus):
+            raise ValueError(f"ExternalResource had invalid status: '{self.status}'")
+
+    @final
+    def activate(self, is_local: bool) -> "ExternalResource":
+        """Perform the initialization of the resource.
+
+        Parameters
+        ----------
+        is_local:
+            True if a local version of the resource should be initialized, False
+            otherwise. If the resource can't be initialized locally, can raise
+            NotImplementedError. If is_local is False, this logic will be executed
+            on the server.
+
+        Returns
+        -------
+        An updated copy of the object
+        """
+        logger.debug("Activating %s", self)
+        updated = self._do_activate(is_local)
+        self._validate_transition(updated)
+        if updated.status.state != ResourceState.ACTIVATING:
+            raise IllegalStateTransitionError(
+                "Calling .activate() should always leave the resource in the "
+                "ACTIVATING state."
+            )
+        return updated
+
+    def _do_activate(self, is_local: bool) -> "ExternalResource":
+        raise NotImplementedError(
+            "Subclasses of ExternalResource should implement ._do_activate(is_local)"
+        )
+
+    @final
+    def deactivate(self) -> "ExternalResource":
+        """Clean up the resource.
+
+        Returns
+        -------
+        An updated copy of the object
+        """
+        logger.debug("Deactivating %s", self)
+        updated = self._do_deactivate()
+        self._validate_transition(updated)
+        if updated.status.state != ResourceState.DEACTIVATING:
+            raise IllegalStateTransitionError(
+                "Calling .deactivate() should always leave the resource in the "
+                "DEACTIVATING state."
+            )
+        return updated
+
+    def _do_deactivate(self) -> "ExternalResource":
+        raise NotImplementedError(
+            "Subclasses of ExternalResource should implement ._do_deactivate()"
+        )
+
+    def _validate_transition(self, updated: "ExternalResource"):
+        if (
+            self.status.state != updated.status.state
+            and not self.status.state.is_allowed_transition(updated.status.state)
+        ):
+            raise IllegalStateTransitionError(
+                f"Cannot transition resource {self.id} from state "
+                f"{self.status.state.name} to state: "
+                f"{updated.status.state.name}"
+            )
+        if self.status.last_update_epoch_time > updated.status.last_update_epoch_time:
+            raise IllegalStateTransitionError(
+                f"Cannot change last_update_epoch_time to a time in the past. "
+                f"Current update time: {self.status.last_update_epoch_time}. New "
+                f"update time: {updated.status.last_update_epoch_time}"
+            )
+
+    @final
+    def update(self) -> "ExternalResource":
```

Comment:
Please add a docstring. My understanding is that this needs to block until a legal transition is performed.